File: Week 1/Problem Set/Credit/credit.c
#include <cs50.h>
#include <stdio.h>

long get_card_number(void);
int digits_counter(long credit_card_number);
int starting_digits(long credit_card_number);
int luhns_algorithm(long credit_card_number);

int main(void)
{
    long credit_card_number = get_card_number();
    int credit_card_length = digits_counter(credit_card_number);
    int first_digits = starting_digits(credit_card_number);
    int luhns = luhns_algorithm(credit_card_number);

    if (luhns % 10  == 0 && credit_card_length == 15 && (first_digits == 34 || first_digits == 37))
    {
        printf("AMEX\n");
        return 0;
    }
    if (luhns % 10  == 0 && credit_card_length == 16 && (first_digits == 51 || first_digits == 52 || first_digits == 53
            || first_digits == 54 || first_digits == 55))
    {
        printf("MASTERCARD\n");
        return 0;
    }
    if (luhns % 10  == 0 && (credit_card_length == 13 || credit_card_length == 16) && (int(first_digits) / 10 == 4))
    {
        printf("VISA\n");
        return 0;
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }

}

long get_card_number(void)
{
    unsigned long credit_card_number = 0;
    do
    {
        credit_card_number = get_long("Number: "); // Prompt for card number repeatedly while it's equal to 0, since it's and unsigned long
    } // If we get past the max value, it will be set at 0 and prompt the user again
    while (credit_card_number == 0);

    return credit_card_number;

}
int digits_counter(long credit_card_number)
{
    int i = 0;
    while (credit_card_number > 0)
    {
        credit_card_number = credit_card_number / 10; // Divide the card number by 10 while it's bigger than 0
        i++; // Each time we divide it by 10 we take the last number out
    }

    return i;
}
int starting_digits(long credit_card_number)
{
    do
    {
        credit_card_number = credit_card_number / 10; // Divide the card number until it's smaller than 100
    }
    while (credit_card_number > 100);

    return credit_card_number;
}

int luhns_algorithm(long credit_card_number)
{
    int division_counter = 0;
    int odd_sum = 0;
    int even_sum = 0;

    while (credit_card_number > 0)
    {
        int last_digit = credit_card_number % 10;
        credit_card_number = credit_card_number / 10;
        division_counter++;

        if (division_counter % 2 != 0) // Every odd number will be added to this sum
        {
            odd_sum = odd_sum + last_digit;
        }
        if (division_counter % 2 == 0) // Every even number will be added to this sum
        {
            last_digit = last_digit * 2; // Multiply the last digit by 2

            if (last_digit <= 9) // if the number is smaller than 9 will immediately be added
            {
                even_sum = even_sum + last_digit;
            }
            if (last_digit > 9) // if it's bigger than 9 will be separated into 2 digits and added to each other
            {
                int first_digit = last_digit % 10;
                int second_digit = last_digit / 10;

                even_sum = even_sum + (first_digit + second_digit);
            }
        }
    }
    return odd_sum + even_sum;
}

File: Week 6/2. Problem Set/4. Credit/credit.py
from cs50 import get_int
# In order for this to work we have to do casting in a bunch of variables
# Since python don't let you choose the datatype beforehand
# Besides this, it's basically the same thing we did in Week 1
# Just using the Python semantic


def main():
    credit_card_number = get_credit_card_number()
    amount_of_digits = get_amount_of_digits(credit_card_number)
    starting_digits = int(get_starting_digits(credit_card_number))
    luhns = int(luhns_algorithm(credit_card_number))
    first_digit = starting_digits / 10
    first_digit = int(first_digit)

    if (luhns % 10 == 0 and amount_of_digits == 16 and (starting_digits == 51 or starting_digits == 52 or starting_digits == 53 or starting_digits == 54 or starting_digits == 55)):
        print("MASTERCARD")
    elif (luhns % 10 == 0 and amount_of_digits == 15 and (starting_digits == 34 or starting_digits == 37)):
        print("AMEX")
    elif (luhns % 10 == 0 and (amount_of_digits == 13 or amount_of_digits == 16) and (first_digit == 4)):
        print("VISA")
    else:
        print("INVALID")


def get_credit_card_number():
    while True:
        credit_card_number = get_int("Number: ")
        if credit_card_number > 100:
            break
    return credit_card_number


def get_amount_of_digits(credit_card_number):

    i = -1
    while credit_card_number > 0:
        credit_card_number = int(credit_card_number) / 10
        i += 1

    return i


def get_starting_digits(credit_card_number):

    while credit_card_number > 100:
        credit_card_number = int(credit_card_number) / 10

    return credit_card_number


def luhns_algorithm(credit_card_number):

    division_counter = 0
    odd_sum = 0
    even_sum = 0

    while (credit_card_number > 0):
        last_digit = int(credit_card_number) % 10
        credit_card_number = int(credit_card_number) / 10
        credit_card_number = int(credit_card_number)
        division_counter += 1

        if (division_counter % 2 != 0):
            # Every odd number will be added to this sum
            odd_sum = odd_sum + last_digit

        if (division_counter % 2 == 0):
            # Every even number will be added to this sum
            last_digit = last_digit * 2
            # Multiply the last digit by 2

            if (last_digit <= 9):
                # if the number is smaller than 9 will immediately be added
                even_sum = even_sum + last_digit

            if (last_digit > 9):
                # if it's bigger than 9 will be separated into 2 digits and added to each other
                first_digit = last_digit % 10
                second_digit = last_digit / 10

                even_sum = even_sum + (int(first_digit) + int(second_digit))

    return odd_sum + even_sum


if __name__ == "__main__":
    main()